daemon/firewall/firewall_manager.go

```diff
@@ -39,10 +56,111 @@ func NewFirewallManager(devices device.ListFunc,
 	ip6TablesSupported bool,
 	enabled bool) FirewallManager {
 	return FirewallManager{
-		iptablesManager: iptablesmanager.NewIPTablesManager(cmdRunner, ip6TablesSupported, enabled),
-		devices:         devices,
-		connmark:        connmark,
+		iptablesManager:    iptablesmanager.NewIPTablesManager(cmdRunner, enabled, ip6TablesSupported),
+		devices:            devices,
+		allowIncomingRules: make(map[string]meshIncomingRule),
+		fileshareRules:     make(map[string]iptablesmanager.FwRule),
+		connmark:           connmark,
+	}
+}
+
+func (f *FirewallManager) AllowIncoming(peer meshnet.UniqueAddress, allowLocal bool) error {
+	if _, ok := f.allowIncomingRules[peer.UID]; ok {
+		return ErrRuleAlreadyActive
+	}
+
+	blockLANRules := []iptablesmanager.FwRule{}
+	if !allowLocal {
+		lans := []string{
+			"169.254.0.0/16",
+			"192.168.0.0/16",
+			"172.16.0.0/12",
+			"10.0.0.0/8",
+		}
+
+		for _, lan := range lans {
+			rule := iptablesmanager.NewFwRule(
+				iptablesmanager.Input,
+				iptablesmanager.IPv4,
+				fmt.Sprintf("-s %s/32 -d %s -j DROP", peer.Address, lan),
+				MeshnetBlockIncomingLAN)
+			blockLANRules = append(blockLANRules, rule)
+
+			if err := f.iptablesManager.InsertRule(rule); err != nil {
+				return fmt.Errorf("blocking mesh peer from LAN access: %w", err)
+			}
+		}
 	}
+
+	rule := iptablesmanager.NewFwRule(
+		iptablesmanager.Input,
+		iptablesmanager.IPv4,
+		fmt.Sprintf("-s %s/32 -j ACCEPT", peer.Address),
+		MeshnetIncoming)
+
+	if err := f.iptablesManager.InsertRule(rule); err != nil {
+		return fmt.Errorf("allowing incoming traffic for peer: %w", err)
+	}
+
+	f.allowIncomingRules[peer.UID] = meshIncomingRule{
```

Comment:
What would happen if one of the `InsertRule` will fail until reaching here, so some of the rules will be set but not saved in the list? Will it be possible to remove the already added rules?

Reply:
For now the rules will remain. Depending on how the new layer is integrated, it might be possible to remove them. I want to add some sort of cleanup on failure after all the kinds of rules are implemented(to generalize it if possible).